src/dfmpreview/src/extension.ts

```diff
@@ -0,0 +1,306 @@
+		//console.log(doc.fileName);
+		let filename = doc.fileName;
+		let rtpath_length = rtpath.length;
+		let filepath = filename.substr(rtpath_length+1 , filename.length - rtpath_length);
+
+		if(doc.languageId=== "markdown"){
+			let num_of_row = docContent.split("\r\n").length;
+			this._spawn.stdin.write(num_of_row + '\n');
+			this._spawn.stdin.write(rtpath + '\n');
+			this._spawn.stdin.write(filepath + '\n');
+			this._spawn.stdin.write(docContent + '\n');
+		}
+	}
+
+    public callDfm(){
+		if(this._is_firsttime){
+			//for the firt time , because the activeTextEditor wil be translate to the viewColumn.two.
+			this._is_firsttime = false;
+			this.sendtext();
+		}
+		else if (!this._waiting) {
+			this._waiting = true;
+			setTimeout(() => {
+				this._waiting = false;
+				this.sendtext();
+			}, 300);
+		}
+    }    
+}
+
+
+class MDDocumentContentProvider implements TextDocumentContentProvider {
+	private _context: ExtensionContext;
+	private _onDidChange = new EventEmitter<Uri>();
+	private _waiting : boolean;
+    
+
+	constructor(context: ExtensionContext) {
+		this._context = context;
+		this._waiting = false;
+	}
+
+	private getMediaPath(mediaFile) : string {
+		return this._context.asAbsolutePath(path.join('media', mediaFile));
+	}
+
+	public provideTextDocumentContent(uri: Uri): Thenable<string> {
+
+		return workspace.openTextDocument(Uri.parse(uri.query)).then(document => {
+			const head = [].concat(
+				'<!DOCTYPE html>',
+				'<html>',
+				'<head>',
+				'<meta http-equiv="Content-type" content="text/html;charset=UTF-8">',
+				//`<link rel="stylesheet" type="text/css" href="${this.getMediaPath('docfx.vendor.css')}" >`,
+				//`<link rel="stylesheet" type="text/css" href="${this.getMediaPath('docfx.css')}" >`,
+				`<link rel="stylesheet" type="text/css" href="${this.getMediaPath('tomorrow.css')}" >`,
+				`<link rel="stylesheet" type="text/css" href="${this.getMediaPath('markdown.css')}" >`,
+				`<link rel="stylesheet" type="text/css" href="${this.getMediaPath('main.css')}" >`,
+				//this.computeCustomStyleSheetIncludes(uri),
+				`<base href="${document.uri.toString(true)}">`,
+				'</head>',
+				'<body>'
+			).join('\n');
+
+            const body = previewresult;
+
+			const tail = [
+				`<script type="text/javascript" src="${this.getMediaPath('docfx.vendor.js')}"></script>`,
+				//`<script type="text/javascript" src="${this.getMediaPath('docfx.js')}"></script>`,
+				`<script type="text/javascript" src="${this.getMediaPath('main.js')}"></script>`,
+				`<script>hljs.initHighlightingOnLoad();</script>`,
+				'</body>',
+				'</html>'
+			].join('\n');
+
+			return head + body + tail;
+		});
+	}
+
+	get onDidChange(): Event<Uri> {
+		return this._onDidChange.event;
+	}
+
+	public update(uri: Uri) {/*
+		if (!this._waiting) {
+			this._waiting = true;
+			setTimeout(() => {
+				this._waiting = false;
+				this._onDidChange.fire(uri);
+			}, 500);
+		}*/
+		this._onDidChange.fire(uri);
+	}
+}
```

Comment:
format the code with 'Alt+Shift+F', and make sure remove leading white spaces, use whitespaces, remove continuous blank lines